faunadb/functions_auth.go

```diff
@@ -0,0 +1,99 @@
+package faunadb
+
+// Authentication
+
+// Login creates a token for the provided ref.
+//
+// Parameters:
+//  ref Ref - A reference with credentials to authenticate against.
+//  params Object - An object of parameters to pass to the login function
+//    - password: The password used to login
+//
+// Returns:
+//  Key - a key with the secret to login.
+//
+// See: https://app.fauna.com/documentation/reference/queryapi#authentication
+func Login(ref, params interface{}) Expr {
+	return loginFn{Login: wrap(ref), Params: wrap(params)}
+}
+
+type loginFn struct {
+	fnApply
+	Login  Expr `json:"login"`
+	Params Expr `json:"params"`
+}
+
+func (fn loginFn) String() string { return printFn(fn) }
```

Comment:
You can drop all `String()` functions and implement it in the `fnApply` struct only.

Reply:
I tried this and it only retrieves fields exported in fnApply. I also tried changing fnApply to an interface but then I can't implement String() on an interface.

Reply:
Okay. I know we can reflect on embedded fields from the outer scope. I assumed the opposite was also true. I apologize for misleading you.

In that case, does the `fnApply` serve us any propose or can we drop it?

Reply:
At the very least fnApply prevents having to add a json ignore field tag to every struct since we embed the `Expr` type  
```go
type atFn struct {
	Expr     `json:"-"`
	At         Expr `json:"at"`
	Expression Expr `json:"expr"`
}
```

It also has no effect since it is not exported.

Reply:
I think this is incidental to the usage of `fnApply` as an embedded struct. Nevertheless, it buys us the implementation of `fn expr()`. I think we should keep it.